sympy/combinatorics/rewritingsystem.py

```diff
@@ -285,3 +289,236 @@ def reduce(self, word, exclude=None):
+        generators = list(self.alphabet)
+        generators += [gen**-1 for gen in generators]
+
+        # The symbols present in the new rules are the symbols to be verified at each state.
+        # computes the automaton_alphabet, as the transitions solely depend upon the new states.
+        for rule in rules:
+            automaton_alphabet += rule.letter_form_elm
+        automaton_alphabet = set(automaton_alphabet)
+
+        # Store the complete left hand side of the rules - dead states.
+        left_hand_rules = list(all_rules)
+
+        # Compute the proper prefixes for every rule.
+        for r in all_rules:
+            proper_prefixes[r] = []
+            letter_word_array = [s for s in r.letter_form_elm]
+            for i in range (1, len(letter_word_array)):
+                letter_word_array[i] = letter_word_array[i-1]*letter_word_array[i]
+            proper_prefixes[r] = letter_word_array
+
+        # Add dead states.
+        for rule in left_hand_rules:
+            if not rule in self.reduction_automaton.states:
+                self.reduction_automaton.add_state(rule, is_dead=True)
+
+        # Add accept states.
+        for r in all_rules:
+            prop_prefix = proper_prefixes[r]
+            for elem in prop_prefix:
+                if not elem in self.reduction_automaton.states:
+                    self.reduction_automaton.add_state(elem, is_accept=True)
+
+        # Add new transitions for every state.
+        for state in self.reduction_automaton.states:
+            current_state_name = state
+            current_state_type = self.reduction_automaton.states[state].state_type
+            if current_state_type == "start":
+                for letter in automaton_alphabet:
+                    if letter in self.reduction_automaton.states:
+                        self.reduction_automaton.states[state].add_transition(letter, letter)
+                    else:
+                        self.reduction_automaton.states[state].add_transition(letter, current_state_name)
+            if current_state_type == "accept":
+                for letter in automaton_alphabet:
+                    next = current_state_name*letter
+                    len_next_word = len(next)
+                    while True:
+                        if len(next) <= 1:
+                            if next in self.reduction_automaton.states:
+                                self.reduction_automaton.states[state].add_transition(letter, next)
+                            else:
+                                self.reduction_automaton.states[state].add_transition(letter, 'start')
+                            break
+                        else:
+                            if next in self.reduction_automaton.states:
+                                self.reduction_automaton.states[state].add_transition(letter, next)
+                                break
+                            next = next.subword(1, len(next))
```

Comment:
All of the above `for` loop could be done more efficiently - otherwise, you might just as well construct the whole automaton again and there is no point for this function. You only need to add transitions that lead to new rules, i.e. to something in `proper_prefixes`. So you can skip anything that obviously can't lead to anything in `proper_prefixes`. E.g. you can't go from a state `a*c**2` to `a*b` because `c` doesn't even occur in `a*b`. In general, if you have a state of the form `a_1*a_2*...*a_n`, you can have a transition from it to `b_1*b_2*...*b_k` if and only if `a_n == b_j` for some `j` and for some such `j` `a_(n-j)*...*a_n` == `b_1*b_2*...*b_j`. There can be many ways to check this. I think, I would go for keeping a dictionary with elements of `alphabet` as keys and a dictionary of the form `{state_name: l}` where `state_name` is something in `proper_prefixes` and `l` are all of the positions `j` s.t. the `j`-th generator of `state_name` is equal to the key. Then for each given state and letter it'll be easy to check if any transitions from it to something in `proper_prefixes` is ever possible. Does that makes sense?

Reply:
If I'm not wrong, I don't think there is a need to check the transitions at the previous states. For example, if the left-hand side of the newly added rule is `a*b**2` then the new states that will be added to the automaton are `a`, `a*b`, `a*b**2` with the last one being a dead state. The only way to reach `a*b**2` is through `a*b` and `a*b` is through `a`, which already belongs to the `proper_prefixes` of the new rules. 

So, I think it would suffice if we just add the transitions to the newly added states and modify the transitions of the `start` state, right?

Reply:
I don't think that's quite the case. What if you already have a state `a*b` that previously led to `a*b*c` on reading `c` but now can also lead to `a*b**2` on reading `b`? Or, for instance, `c*a*b` that leads to `c*a*b*a` on reading `a` but now could go directly to `a*b**2` if it reads `b`?

Reply:
Yeah, Thank you. Makes sense.